src/test/crimson/test_messenger.cc

```diff
@@ -572,6 +578,1650 @@ seastar::future<> test_preemptive_shutdown(bool v2) {
+    return cmd_msgr.start(this).then([this, cmd_peer_addr] {
+      return cmd_msgr.connect(cmd_peer_addr, entity_name_t::TYPE_OSD);
+    }).then([this] (auto conn) {
+      cmd_conn = conn->release();
+      return pingpong();
+    });
+  }
+
+ public:
+  FailoverTest(Messenger& cmd_msgr,
+               entity_addr_t test_addr,
+               entity_addr_t test_peer_addr)
+    : cmd_msgr(cmd_msgr),
+      test_addr(test_addr),
+      test_peer_addr(test_peer_addr) { }
+
+  seastar::future<> shutdown() {
+    logger().info("CmdCli shutdown...");
+    assert(!recv_cmdreply);
+    MCommand *m = new MCommand();
+    m->cmd.emplace_back(1, static_cast<char>(cmd_t::shutdown));
+    return cmd_conn->send(MessageRef{m, false}).then([this] {
+      return seastar::sleep(200ms);
+    }).finally([this] {
+      return cmd_msgr.shutdown();
+    });
+  }
+
+  static seastar::future<seastar::lw_shared_ptr<FailoverTest>>
+  create(entity_addr_t cmd_peer_addr, entity_addr_t test_addr) {
+    assert(cmd_peer_addr.is_msgr2());
+    return Messenger::create(entity_name_t::OSD(1), "CmdCli", 1, 0
+    ).then([cmd_peer_addr, test_addr] (Messenger* cmd_msgr) {
+      entity_addr_t test_peer_addr = cmd_peer_addr;
+      test_peer_addr.set_port(cmd_peer_addr.get_port() + 1);
+      test_peer_addr.set_nonce(4);
+      auto test = seastar::make_lw_shared<FailoverTest>(
+          *cmd_msgr, test_addr, test_peer_addr);
+      return test->init(cmd_peer_addr).then([test] () mutable {
+        logger().info("CmdCli ready");
+        return test;
+      });
+    });
+  }
+
+ // called by tests
+ public:
+  seastar::future<> run_suite(std::string name,
+                              const TestInterceptor& interceptor,
+                              policy_t test_policy,
+                              policy_t peer_policy,
+                              std::function<seastar::future<>(FailoverSuite&)>&& f) {
+    logger().info("\n\n[{}]", name);
+    ceph_assert(!test_suite);
+    SocketPolicy test_policy_ = toSocketPolicy(test_policy);
+    return FailoverSuite::create(test_addr, test_policy_, test_peer_addr, interceptor
+    ).then([this, peer_policy, f = std::move(f)] (auto suite) mutable {
+      test_suite.swap(suite);
+      return start_peer(peer_policy).then([this, f = std::move(f)] () mutable {
+        return f(*test_suite);
+      }).then([this] {
+        test_suite->post_check();
+        logger().info("\n[SUCCESS]");
+      }).handle_exception([] (auto eptr) {
+        logger().info("\n[FAIL: {}]", eptr);
+        throw;
+      }).finally([this] {
+        return stop_peer();
+      }).finally([this] {
+        return test_suite->shutdown().then([this] {
+          test_suite.reset(nullptr);
+        });
+      });
+    });
+  }
+
+  seastar::future<> peer_connect_me() {
+    return prepare_cmd(cmd_t::suite_connect_me,
+        [this] (MCommand* m) {
+      ostringstream sout;
+      sout << test_addr;
+      auto str_test_addr = sout.str();
```

Comment:
you could use `fmt::format()` as a better `printf()`, like:
```c++
m->cmd.emplace_back(fmt::format("{}", test_addr));
```

so no need to use `ostringstream` when building a command string or an exception.

Reply:
Good to know, all fixed.